server/src/main/java/com/thoughtworks/go/server/dao/DataSharingSettingsSqlMapDao.java

```diff
@@ -48,11 +55,32 @@ public void saveOrUpdate(DataSharingSettings dataSharingSettings) throws Duplica
             existing = dataSharingSettings;
         }
 
-        sessionFactory.getCurrentSession().saveOrUpdate(existing);
+        DataSharingSettings toSave = existing;
+        transactionTemplate.execute(new TransactionCallbackWithoutResult() {
+            @Override
+            protected void doInTransactionWithoutResult(TransactionStatus status) {
+                transactionSynchronizationManager.registerSynchronization(new TransactionSynchronizationAdapter() {
+                    @Override
+                    public void afterCommit() {
+                        invalidateCache();
+                    }
+                });
+
+                sessionFactory.getCurrentSession().saveOrUpdate(toSave);
+            }
+        });
     }
 
     public DataSharingSettings load() {
-        return transactionTemplate.execute(status -> (DataSharingSettings) sessionFactory.getCurrentSession().getNamedQuery("load.datasharing.settings").uniqueResult());
+        if (cachedSettings == null) {
+            cachedSettings = transactionTemplate.execute(status -> (DataSharingSettings) sessionFactory.getCurrentSession().getNamedQuery("load.datasharing.settings").uniqueResult());
+        }
+
+        return cachedSettings;
+    }
+
+    public void invalidateCache() {
```

Comment:
Also, this method need not be public.

Reply:
I know. but the integration Test setup is in a manner that `DBHelper.onTeardown()` will clear off the data from the DB but won't clear the cached object. To do so, we need it public.

Another workaround would be to put it under `GoCache`, but the cached object is accessed within that specific DAO instance, hence kept the cached object local and exposed a method to clear it off!

Reply:
I think we should stick to `GoCache`, its not a workaround.

Reply:
Ok